cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,220 @@
+
+CIDs are presenting issues with reliability and scalability, and are causing network policies to be in a broken state. Most notable cases:
+
+* Reaching maximum number (65k) of CIDs in the cluster.
+* eBPF policy maps overflowing with over 16k entries.
+
+# Design
+
+## Operator changes
+
+Operator will start managing CIDs, and will become the only system pod that issues mutating requests for CIDs to the kube-apiserver. A new controller (CID controller) will run inside Operator that will be responsible for managing (creating) CIDs.
+
+1) Operator requires watching the following resources to allocate CIDs:
+
+* Pods (Already watches) - To manage CIDs based on pod labels.
+* Namespaces - to manage CIDs based on namespace labels changes.
+* CIDs - to keep the desired state of CIDs. Required for:
+    * Getting the initial state on start-up.
+    * Being compatible with Agents managing CIDs simultaneously.
+    * Handling unplanned CID changes (create, update, delete) by other sources -- not Operator or Agent.
+
+2) Operator will modify the following resources:
+
+* CIDs - Create and update. If namespace or pod labels are changed, new CIDs will be created for all unique label sets that don’t match any existing identity. Unused CIDs will be cleaned up by the CID GC. In case of an update to a CID, the CID controller will update the CID back to the correct state, effectively reverting the initial change.
+
+3) CID GC:
+
+* CID garbage collection continues to be the responsibility of the Operator. In the initial release, this will still rely on the legacy CID GC (the periodic GC) to deleted unused CIDs. The design of this can be evaluated after completing this CFP.
+
+## Agent changes
+
+Agents will stop creating CIDs. Agents will only generate local security identities. Agents will no longer perform following actions:
+
+* Locally create and reserve security identities in the global range [256, 65536].
+* Create CIDs.
+
+Subsequent milestones will further remove the following from the Agents:
+* Reporting heartbeat status annotation on CIDs.
+* Permissions to modify CIDs.
+
+Agents will wait for CIDs to be received through the watcher, and then assign them to local endpoints (pods) by retrieving the CIDs from the watcher's store based on pod and namespace labels.
+
+**Namespace watch**
+
+Agents need to know what are the namespace labels for each pod, to be able to match pod+namespace labels to a CID from CID watcher's cache. Agents will continue to watch namespaces in order to instantly update security identity labels in the local Endpoint manager’s cache, and start attempting to get CID from CID watcher store, to update the corresponding endpoint. More info below in Label changes: Namespace.
+
+**Pod startup**
+
+CNI will be blocked if CID doesn't exist in the CID watcher's store. After retries, CNI will fail and pod creation will fail. This is useful to quickly identify issues with CID controller or Operator.
```

Comment:
If this is a strong pain point, I think we can explore improvements in this area as a subsequent milestone. Early versions of the CFP had a lot of complexity to try to solve this problem and we can't even quantify it yet because the overall architecture is not in place.

I would also challenge the idea that Security Identities are constantly changing. If Security Identities are relatively static or preallocated then pod startup time should not be heavily affected.

Reply:
pod startup time is critical metric for us and wanted to see if it has any real impact.

> we can't even quantify it yet because the overall architecture is not in place.

Agreed

Reply:
Pod startup latency can be increased, but only if there is an issue with Operator.
Otherwise, pod scheduling depends on kube-apiserver being healthy, and if kube-apiserver is healthy, Operator will be able to create CID before the pod even gets scheduled (Operator creates CIDs on Pod object creation).

This means that pod startup latency will be decreased, because Agents will no longer need to create CID (request to kube-apiserver), but just read it locally from the watcher's store.

Keep in mind, the difference is only when a CID doesn't exist for the label set for a new pod.

Reply:
> Pod startup latency can be increased, but only if there is an issue with Operator.

Thanks @dlapcevic. I think it depends on how much operator can scale. Operator can become single point of failure which may potentially delay pod creation in every node. I have one follow-up question. Does this model allow multiple cilium operator working together (like active-active) or its more of primary/secondary model (like replicas used only as backups)?

Reply:
i guess former otherwise it would again endup in duplicate identity creation. this is a trade-off for going with centralized management